### ISSUE TITLE
Default "Fill" option inappropriate for unused address fields

### ISSUE THREAD

**Issue by twomice** [ginkgostreet/com.ginkgostreet.geofill#7]
Testing scenario:

1. On the "Settings - Addresses" page (/civicrm/admin/setting/preferences/address?reset=1), field collection "Address Editing:" "Address Fields", the "Postal Code Suffix" checkbox is unchecked. So "Postal Code Suffix" field is never editable in the UI.
2. Install this extension and leave its settings in their default ("Fill" for all address components.)
3. Add an address that would have a zip+4 component and observe that the postal-code suffix is displayed with the address.
4. Observe that it's impossible to edit that postal-code suffix for this address manually.
5. Also observe that any further edits to this address will retain that postal-code suffix value, even if it's incorrect.

Recommendation:
Upon installation: Set set "Geodata Filler Extension Settings" to "Discard" by default for all address components, but set to "Fill" the ones which are enabled for editing.

**Comment by universalhandle** [comment]
I didn't reproduce your steps, but I _am_ picking up what you're putting down. I think there are actually a few different problems here:

1. Ideally we'd have some extra logic in the extension to adapt to the address settings an instance has configured. For instance, if the county field is disabled, it doesn't make a lot of sense to set a policy for it. Pull request #5 is another example of this; the contributed code adds support for parsing the street_name, street_number, etc., but arguably the extension should do this only if the administrator has turned on  address settings are configured to allow that. While desirable, this level of sophistication is more than was necessary for my initial use case and I'm not likely to add it without funding.
2. The "Fill" setting is potentially problematic. Suppose a contact enters the database with a Florida address, undergoes geocoding, and has the county, zip-extension, etc. populated. If user then moves to New York, the county, the zip-extension, etc. will become stale unless manually updated. I'm not sure what to do about this. There might be edge cases when you want to "Fill" only (e.g., during an import of new contacts), but it's clearly better for every-day usage if the user chooses "Discard" or "Overwrite." I could remove the "Fill" option, or I could throw a system warning whenever the option is set (to remind the user that their site is configured for an edge case). I guess of these two ideas, the latter is more attractive to me. Any other thoughts?
3. The extension should ship with the most conservative possible configuration, which is to do nothing (i.e., preserve core behavior). That's an easy enough fix.

**Comment by joemcl** [comment]
Yeah I've tried editing a ZIP+4 code added by the extension. I deleted the ZIP+4 value for my Home address, saved that edit, but the ZIP+4 re-appears, was not deleted. I'm on 4.7.20 .

**Comment by universalhandle** [comment]
@joemcl: Well in that case, the extension is doing exactly what you've configured it to: fill in the missing data. Nothing wrong with that. I'm more concerned about what happens when you put in a completely new address.

**Comment by joemcl** [comment]
@GinkgoFJG OK so I edited my Home address to a new and different address. The correct ZIP+4 was added, and the record does not show the old ZIP+4. Noting that I don' t see the Suffix box to edit it (probably a Display setting I need to change). The ZIP+4 shows in Postal Code - but only the first five digits show when I edit the address.
